File: Excpute-JS/Display.js
const { ram_read, ram_write } = require("./RAM");

function init_display(scale) {
  const canvas = document.getElementById("display");
  const ctx = canvas.getContext("2d");

  const width = 256;
  const height = 256;

  const dpr = window.devicePixelRatio || 1;

  canvas.width = width * dpr * scale;
  canvas.height = height * dpr * scale;

  canvas.style.width = `${width * scale}px`;
  canvas.style.height = `${height * scale}px`;

  ctx.scale(dpr * scale, dpr * scale);
}

function update_display() {
  const x = ram_read(253, false);
  const y = 255 - ram_read(254, false);
  const r = ram_read(250, false);
  const g = ram_read(251, false);
  const b = ram_read(252, false);
  const mode = ram_read(255, false);

  if (mode == 1) {
    // Set pixel
    ctx.fillStyle = color;
    ctx.fillRect(x, y, 1, 1);
  } else if (mode == 2) {
    // Reset pixel
    ctx.fillStyle = "#000000";
    ctx.fillRect(x, y, 1, 1);
  } else if (mode == 3) {
    // Fill screen
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, 255, 255);
  }

  ram_write(255, 0, true);
}

module.exports = { init_display, update_display };

File: Excpute-PY/Display.py
import pygame
import RAM


def init_display(scale):
    resolutionX = 256
    resolutionY = 256
    width = resolutionX * scale
    height = resolutionY * scale

    pygame.init()
    screen = pygame.display.set_mode((width, height))
    pygame.mouse.set_visible(False)
    # clock = pygame.time.Clock()
    return screen



def update_display(screen, scale):
    x = RAM.read(253, False)
    y = 255 - RAM.read(254, False)
    r = RAM.read(250, False)
    g = RAM.read(251, False)
    b = RAM.read(252, False)
    mode = RAM.read(255)
    
    if mode == 1:
        # print("set color")
        pixel_color = (r, g, b)
        pixel_rect = pygame.Rect(x * scale, y * scale, scale, scale)
        pygame.draw.rect(screen, pixel_color, pixel_rect)
    elif mode == 2:
        # print("fill color")
        pixel_color = (r, g, b)
        screen.fill(pixel_color)
    elif mode == 3:
        # print("update buffer")
        pygame.display.flip()
    RAM.write(255, 0, False)